### ISSUE TITLE
"Unauthorized" Response with Hook Data

### ISSUE THREAD

**Issue by cvalarida** [feathersjs/feathers-authentication#291]
Bad title, I know. Sorry.

The story goes like this:
I'm implementing authentication on an app that I'm migrating from Laravel. Previously, I did everything manually, so I knew exactly what was going on. Now that I'm using Feathers JS and the feathers-authentication package, there's some magic involved and I don't know what's up now.

After creating the Feathers app via the feathers generator, I tweaked the endpoints of my api calls to prepend them with `/api`--that's the only thing I've done different from the default as far as I can recall. That said, here's the code:

## Client ##

```javascript
///////////////////
// Initial setup //
///////////////////
const app = feathers()
  .configure(hooks())
  .configure(socketio(socket))
  .configure(authentication({ storage: window.localStorage }));

// Attach it to window for global use
window.app = app;



// Later on, when I try to log in...
window.app.authenticate({
	type: 'local',
	email: username,
	password: password
}).then(function (response) {
	console.log("Success:", response);
}).catch(function (error) {
	console.log('Failure:', error);
});
```

## Server ##
Authentication service...
```javascript
//////////////////////////////
// Configure authentication //
//////////////////////////////
let config = app.get('auth');
app.configure(authentication(config));
```

The config in question:
```json
{
  "auth": {
    "idField": "id",
    "token": {
      "secret": "redacted"
    },
    "local": {
      "userEndpoint": "/api/users"
    }
  }
}
```

## The Problem ##
When I send in my authentication request with the wrong password, I get the following response (from examining the websocket):

```json
[
  "unauthorized",
  {
    "name": "NotAuthenticated",
    "message": "Invalid login.",
    "code": 401,
    "className": "not-authenticated",
    "errors": {}
  }
]
```
Looks good, right? Okay, so now let's log in with the proper password...

```json
[
  "unauthorized",
  {
    "hook": {
      "data": {
        "id": 102,
        "accountId": 1,
        "email": "email@mail.com",
        "password": "redacted",
        "api_token": null,
        "permissions": null,
        "deletedAt": null,
        "createdAt": "2016-09-01T00:41:54.725Z",
        "updatedAt": "2016-09-14T17:14:00.401Z"
      },
      "params": {
        "data": {
          "id": 102,
          "accountId": 1,
          "email": "email@mail.com",
          "password": "redacted",
          "api_token": null,
          "permissions": null,
          "deletedAt": null,
          "createdAt": "2016-09-01T00:41:54.725Z",
          "updatedAt": "2016-09-14T17:14:00.401Z"
        }
      },
      "method": "create",
      "type": "after",
      "result": {
        "id": 102,
        "token": "redacted"
      }
    }
  }
]
```
Er...What? Checking my console also shows me `Failure: Object {hook: Object}` so it's clearly running the `.catch()`, but...why??

I'm stuck...Any ideas? Am I doing something wrong? Is it a bug? I'm using `feathers-authentication@0.7.9`.

**Comment by daffl** [comment]
Can you POST with email and password to the `auth/local` route with e.g. Postman?

**Comment by cvalarida** [comment]
I don't think I have REST enabled server-side, so no. I get:
```json
{
  "name": "GeneralError",
  "message": "Cannot read property 'get' of undefined",
  "code": 500,
  "className": "general-error",
  "data": {},
  "errors": {}
}
```
when POSTing
```json
{ "type": "local", "email": "email@mail.com", "password": "newPass" }
```

In other news, I made a repo to demonstrate this issue at https://github.com/cvalarida/feathers-auth-bug

If it's a bit of a pain to set up, sorry--dev ops isn't really my thing. It should be fairly straightforward, though. I've got all the setup instructions in the README.

Thanks.

**Comment by cvalarida** [comment]
Hey! There it is. Finally got it. So dumb... It was a matter of config.

Previously, I had:
```json
{
  "auth": {
    "idField": "id",
    "token": {
      "secret": "redacted"
    },
    "local": {
      "userEndpoint": "/api/users"
    }
  }
}
```
But what I really needed was:
```json
{
  "auth": {
    "idField": "id",
    "token": {
      "secret": "redacted"
    },
    "userEndpoint": "/api/users",
    "local": {}
  }
}
```
So when I called it via REST, it had nothing to get from, since it didn't actually get a user back (had the wrong `userEndpoint`; it was only defined in `local`). Makes much more sense now on that front.

What I don't understand, though, is why it returned an error and the hook data previously (when using the websocket).

Also, could we perhaps put an example in the documentation? It makes loads more sense now, and I see my folly, but I thinks it would make it clearer from the onset if we had a little example.

**Comment by daffl** [comment]
Glad you figured it out! We'll make sure this is more obvious when we release the next version and update the documentation,

**Comment by cvalarida** [comment]
Yeah, after I figured it out, I found [the example](https://docs.feathersjs.com/authentication/readme.html#example-configuration) in the documentation, so I felt pretty brilliant. The only thing I can think of to make it more clear is mentioning using the `userEndpoint` outside of a specific auth provider configuration, otherwise funky stuff happens. Other than that, my problem was just skimming the documentation and putting it in the wrong place despite the example.

Thanks for pointing me in the right direction!
